File: Lecture11/simple_root_finding.py
#!/usr/bin/env python
from math import exp, log

def f (x) :
    return exp(x) * log(x) - x * x;


print ' Simple search for root of exp(x)*log(x) - x*x'
print ' ---------------------------------------------'
x = float (input(' Enter guess x: '))
dx = float (input(' Enter step dx: ') )
acc = float (input(' Enter accuracy: ' ) )


print " Step            x                    dx"
print " ----    ------------------    ------------------"
step = 0
print  ' {0:4.0f}    {1:20.15f} {2:20.15f}'.format(step, x, dx)
f_old = f(x)

while abs(dx) > abs(acc) :
    x += dx
    if f_old * f(x) < 0 :
        x -= dx
        dx /= 2
    
    step += 1
    print  ' {0:4.0f}    {1:20.15f} {2:20.15f}'.format(step, x, dx)

File: Lecture11/simple_root_finding.cpp
#include <cmath>
#include <iostream>
#include <iomanip>
using namespace std;

double f(double x) {
    return exp(x) * log(x) - x * x;
}

void print(int step, double x, double dx) {
    cout.setf(ios::right, ios::adjustfield);
    cout << " " << setw(4) << step << "    ";
    cout.precision(15);
    cout.setf(ios::left, ios::adjustfield);
    cout.setf(ios::showpoint | ios::fixed);
    cout << setw(20) << x << "  " << setw(20) << dx << '\n';
}

int main() {

    cout << " Simple search for root of exp(x)*log(x) - x*x\n"
         << " ---------------------------------------------\n"
         << " Enter guess x, step dx, and desired accuracy: ";
    double x, dx, acc;
    cin >> x >> dx >> acc;

    cout << " Step            x                    dx\n"
         << " ----    ------------------    ------------------\n";
    int step = 0;
    print(step, x, dx);
    double f_old = f(x);
    while (fabs(dx) > abs(acc)) {
        x += dx;
        if ( f_old * f(x) < 0 ) {
            x -= dx;
            dx /= 2;
        }
        ++step;
        print(step, x, dx);
    }
}
